roles/postgresql_upgrade/tasks/main.yml

```diff
@@ -6,40 +6,53 @@
       - ansible_distribution_major_version == '8'
     fail_msg: "This role is only supported on EL8"
 
+- name: Check if PostgreSQL is installed
+  stat:
+    path: /var/lib/pgsql/data/PG_VERSION
+  register: pg_present
+
 - name: Get version from the PostgreSQL data directory
   slurp:
     path: /var/lib/pgsql/data/PG_VERSION
   register: pg_version
+  when: pg_present.stat.exists
 
-- when:
-    - "'10' in pg_version['content']|b64decode"
+- when: >
+    "pg_version is defined and
+    '10' in pg_version['content']|b64decode or
+    not pg_present.stat.exists"
```

Comment:
Can you expand on why these extra conditions are needed?

Reply:
When running this role after the upgrade already has been done, you can run into all kinds of issues (and errors stopping the play).

All these conditions are to make sure that the upgrade will only run if the ```pg_version``` has been set and not the correct version OR when the PG data directory does not exist.

I'll also update the logic to check for the wanted version instead of the assumed lower version, makes more sense

Reply:
The tests for this role should be running an idempotency test (https://github.com/theforeman/foreman-operations-collection/tree/develop/roles/postgresql_upgrade/molecule/redhat) to verify that after it runs once code is not executed again to do any upgrading. What do you think is missing or what scenario is missing in order to ensure the tests catch this?

Reply:
You're right, it does exactly that :-) (I think I got a bit side-tracked working on other things and mixed up a few thoughts). The new feature I added is that the role will now also install postgres if it is unable to find an existing database directory.

The tasks list is here: https://github.com/Thulium-Drake/ansible-role-foreman/blob/master/tasks/install_foreman.yml

Reply:
Why tho? The installer takes care of all things PostgreSQL (besides the upgrade, which this role is for).

Reply:
Should the procedure then be:

1. Install Foreman
2. Upgrade PostgreSQL

In the current playbook I first ensure that PostgreSQL is present and upgraded before running the installer. And in that case, this role goes bust when there is no PostgreSQL installed on the system:

```
TASK [Upgrade postgres on EL8] ***************************************************************************************

TASK [theforeman.operations.postgresql_upgrade : Ensure running on EL8] **********************************************
ok: [deploy.lbhr.htm.lan] => changed=false 
  msg: All assertions passed

TASK [theforeman.operations.postgresql_upgrade : Get version from the PostgreSQL data directory] *********************
fatal: [deploy.lbhr.htm.lan]: FAILED! => changed=false 
  msg: 'file not found: /var/lib/pgsql/data/PG_VERSION'
[WARNING]: Failure using method (v2_runner_on_failed) in callback plugin
(<ansible.plugins.callback.ara_default.CallbackModule object at 0x7fcc409c92e0>): 'NoneType' object is not
subscriptable
```

And with the changes I made to the role it checks if the server is already present, if not, it sets it up

Reply:
This role is only needed when one has a Foreman on EL8 which was installed with PostgreSQL 10 and wants to upgrade to one that expects 12.

By default modern Foreman always uses 12, so this role becomes a noop (we have been including it in our CI unconditionally, without any issues).

But yes, the role is not expected to be executed on empty systems.

Reply:
Right, that makes things a lot clearer! Thank you!  I'll revert the changes I made with regards to the postgresl_upgrade role :)

Reply:
Is there anything in the readme of the role that led you on the wrong path and we should improve instead?

Reply:
Ah, sorry, I'm blind, I posted an answer in the main thead :sweat_smile: 
> @evgeni I have updated the file (which removed the convo it seems). But to answer your question, I read the description of the role as "this will make sure that the EL8 system is running the correct version of psql". So when assembling a playbook that would install Foreman using the roles I assumed I had to run this first before attempting the install